src/utilities/shared_functions.zig

```diff
@@ -182,3 +183,178 @@ pub fn isApproximatelyEqual(value_a: f64, value_b: f64) bool {
+        for (0..nbits) |_| {
+            try bit_writer.writeBits(@as(u1, 0b0), 1);
+        }
+        try bit_writer.writeBits(value, nbits + 1);
+    }
+    try bit_writer.flushBits();
+    return encoded_values;
+}
+
+/// Decodes an array of u8 `compressed_values` previously encoded using Elias Gamma encoding.
+/// The memory `allocator` is used to allocate the resulting decoded data. The function returns
+/// an `ArrayList(u64)` containing the decoded data, or an error if the input is unsupported,
+/// meaning it was not encoded using Elias Gamma encoder or corrupted. Any other memory related
+/// error is also returned.
+pub fn eliasGammaDecoder(allocator: Allocator, compressed_values: []const u8) !ArrayList(u64) {
+    var decoded_values = ArrayList(u64).init(allocator);
+
+    // Create bit reader over full byte slice.
+    var stream = std.io.fixedBufferStream(compressed_values);
+    var bit_reader = std.io.bitReader(.big, stream.reader());
+
+    while (true) {
+        // Count leading zeros.
+        var leading_bits_number: u16 = 0;
+        while (true) {
+            const leading_bit: u8 = bit_reader.readBitsNoEof(u8, 1) catch |err| {
+                if (err == error.EndOfStream) {
+                    // No more values to decode.
+                    return decoded_values;
+                }
+                // Some other error occurred.
+                decoded_values.deinit();
+                return err;
+            };
+
+            if (leading_bit == 1) break;
+            leading_bits_number += 1;
+        }
+
+        // Read the value bits.
+        if (leading_bits_number == 0) {
+            try decoded_values.append(1);
+        } else {
+            const suffix: u64 = bit_reader.readBitsNoEof(u64, leading_bits_number) catch |err| {
+                // If we cannot read all k bits, the stream is malformed.
+                decoded_values.deinit();
+                if (err == error.EndOfStream) {
+                    return Error.ByteStreamError;
+                }
+                return Error.ByteStreamError;
+            };
+            // Combine leading 1 bit with suffix to form the decoded value.
+            const value: u64 = (@as(u64, 1) << @intCast(leading_bits_number)) | suffix;
+            try decoded_values.append(value);
+        }
+    }
+}
+
+test "zigzag can encode and decode signed integers correctly" {
+    const default_random = tester.getDefaultRandomGenerator();
+    const number_of_tests = tester.generateNumberOfValues(default_random);
+    for (0..number_of_tests) |_| {
+        const original = tester.generateBoundRandomInteger(
+            i64,
+            -tester.max_test_value,
+            tester.max_test_value,
+            default_random,
+        );
+        const encoded = zigzagEncoder(original);
+        const decoded = zigzagDecoder(encoded);
+        try testing.expect(decoded == original);
+    }
+}
+
+test "eliasGammaEncoder can encode simple values correctly" {
```

Comment:
Since both `eliasGammaEncoder()` and `eliasGammaDecoder()` has been implemented, maybe only keep _"eliasGammaEncoder cannot encode zero value"_ and _"eliasGammaDecoder can encode and decode complex values correctly"_ as these two tests together should ensure that all `u64` values are encoded correctly and that zero is rejected. Of course, if it is important that 15, 7, 3, 1 are encoded exactly as the test checks, maybe also keep the other two tests.